### PR TITLE
Fix issue 5040 for wires in PauliSentence

### PR DESCRIPTION
**Pull request by Qottmann**
Quick fix for https://github.com/PennyLaneAI/pennylane/issues/5040

### CODE REVIEW

**Comment by Qottmann** [comment]
Noticed that some qchem tests have that behavior baked in, is there a logic behind the (seemingly arbitrary) wire ordering that I am missing 😬 ?

File: pennylane/pauli/pauli_arithmetic.py
[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py:1]: amazing, thanks @soranjh !

**Comment by Qottmann** [comment]
There are some new, non-trivial failures in the tapering tests. I tried to check if it just simple wire re-orderings but there seem to be some genuine discrepancies between what is expected and returned, in particular for the cases where new opmath is enabled 😬 @soranjh tbh its very hard to figure out what is going wrong (and tested in the first place) for me not being deep in the qchem module. Could you (or someone else) take another look at this?

We did no see those earlier because the test suite did not run that "far" in the tests. I ran all qchem tests locally and get that short summary (so main problem seems in the tapering tests indeed)

```
======================================================== short test summary info ========================================================
FAILED tests/qchem/test_tapering.py::test_optimal_sector[symbols0-geometry0-0-generators0-2-result0] - assert [-1, 1, -1] == [1, -1, -1]
FAILED tests/qchem/test_tapering.py::test_optimal_sector[symbols2-geometry2-1-generators2-2-result2] - assert [1, 1] == [-1, -1]
FAILED tests/qchem/test_tapering.py::test_transform_hf[generators1-paulixops1-paulix_sector1-2-6-result1] - assert tensor(False, requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_transform_hf[generators2-paulixops2-paulix_sector2-4-12-result2] - assert tensor(False, requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_obs[symbols0-geometry0-0-True] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_obs[symbols1-geometry1-1-False] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_obs[symbols1-geometry1-1-True] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_obs[symbols2-geometry2-1-False] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_obs[symbols2-geometry2-1-True] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_obs[symbols3-geometry3-0-False] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_obs[symbols3-geometry3-0-True] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_excitations[symbols1-geometry1-1-generators1-paulixops1-paulix_sector1-num_commuting1] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_taper_excitations[symbols2-geometry2-1-generators2-paulixops2-paulix_sector2-num_commuting2] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_consistent_taper_ops[operation2-op_gen2] - assert tensor([[False]], requires_grad=True)
FAILED tests/qchem/test_tapering.py::test_consistent_taper_ops[operation3-op_gen3] - assert False
======================================== 15 failed, 299 passed, 86 skipped, 1 xfailed in 29.24s =========================================
```

**Comment by soranjh** [comment]
> There are some new, non-trivial failures in the tapering tests. I tried to check if it just simple wire re-orderings but there seem to be some genuine discrepancies between what is expected and returned, in particular for the cases where new opmath is enabled 😬 @soranjh tbh its very hard to figure out what is going wrong (and tested in the first place) for me not being deep in the qchem module. Could you (or someone else) take another look at this?
> 
> We did no see those earlier because the test suite did not run that "far" in the tests. I ran all qchem tests locally and get that short summary (so main problem seems in the tapering tests indeed)
> 

There are two places in tapering that assume ascending wire orders:

1. The wiremap [here](https://github.com/PennyLaneAI/pennylane/blob/5cfdeddbb59a7147146a0c550b8e1099de0cc022/pennylane/qchem/tapering.py#L302). To fix the issue, you can just sort the wires:

wiremap = dict(zip(list(wireset.toset()), range(len(wireset) + 1)))

2. The wire check [here](https://github.com/PennyLaneAI/pennylane/blob/5cfdeddbb59a7147146a0c550b8e1099de0cc022/pennylane/qchem/tapering.py#L438). This can be fixed in the same way with:

symmstr = np.array([1 if wire in tau.wires else 0 for wire in qubit_op.wires.toset()])

I believe these changes are enough to pass the tests and I do not think they will cause any issues elsewhere. But we should be careful.

File: pennylane/qchem/tapering.py
Comment (timmysilv) [inline pennylane/qchem/tapering.py:1]: are these `toset()` calls necessary? or rather, does `list(wires.toset())` provide some double-sorting that isn't visibly obvious? it's not clear to me why it's necessary which kinda concerns me; I figured `__iter__` should work the same way

**Review by timmysilv — COMMENTED** [review]
nice fix! glad it works, bit worried about the details but i'm fine with it given we know why

File: pennylane/qchem/tapering.py
Comment (Qottmann) [reply pennylane/qchem/tapering.py:1]: I must admit I took this over as suggested by @soranjh without checking the details

Comment (soranjh) [reply pennylane/qchem/tapering.py:1]: The tapering module assumes sorted wires, as is the case for `Hamiltonian`. The new change arranges wires in `Sum` based on the order they are added. So, sorting them internally is a quick solution.

**Review by timmysilv — APPROVED** [review]
got it, I thought sets sort by entry-order but that is not the case! lgtm 🥾 🐛

**Review by albi3ro — COMMENTED** [review]
Mind adding a test that would fail under master to assert that the bug is "fixed"?

**Comment by Qottmann** [comment]
> Mind adding a test that would fail under master to assert that the bug is "fixed"?

Good point, thanks! Done ✅

File: pennylane/measurements/measurements.py [hunk 1]
```diff
@@ -171,7 +171,8 @@ def __init__(
         self.id = id
 
         if wires is not None:
-            if len(wires) == 0:
+            print(self.obs, type(self), self.__dict__)
```
Comment (soranjh) [inline pennylane/measurements/measurements.py:174]: Please remove this print.

Comment (Qottmann) [reply pennylane/measurements/measurements.py:174]: This was from another PR; no clue how this sneaked in here 😬 thanks for catching it!

File: tests/test_hermitian_edge_cases.py [hunk 2]
```diff
@@ -20,6 +20,8 @@
 
 import pennylane as qml
 
+np.random.seed(0)
```
Comment (Qottmann) [inline tests/test_hermitian_edge_cases.py:23]: same as https://github.com/PennyLaneAI/pennylane/pull/5041/files#r1471023104

**Review by soranjh — APPROVED** [review]

**Pull request merged**